File: src/142.linked-list-cycle-ii/solution_hash-table.ts
/*
 * @lc app=leetcode.cn id=142 lang=typescript
 *
 * [142] 环形链表 II
 */

/**
 * Definition for singly-linked list.
 */
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

// @lc code=start
function detectCycle(head: ListNode | null): ListNode | null {
  if (!head) {
    return null;
  }

  const set = new Set<ListNode>();
  let current: ListNode | null = head;

  while (current) {
    if (set.has(current)) {
      return current;
    }

    set.add(current);
    current = current.next;
  }

  return null;
}
// @lc code=end

export default detectCycle;

File: src/142.linked-list-cycle-ii/solution_hash-table.cpp
/*
 * @lc app=leetcode.cn id=142 lang=cpp
 *
 * [142] 环形链表 II
 */

#include <unordered_set>

using namespace std;

struct ListNode {
  int val;
  ListNode *next;
};

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *   int val;
 *   ListNode *next;
 *   ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
 public:
  ListNode *detectCycle(ListNode *head) {
    if (!head) {
      return nullptr;
    }

    unordered_set<ListNode *> dict;
    ListNode *curr = head;

    while (curr) {
      if (dict.find(curr) != dict.end()) {
        return curr;
      }

      dict.insert(curr);
      curr = curr->next;
    }

    return nullptr;
  }
};
// @lc code=end
